### PR TITLE
SCA: use 'yield from' where it does make sense

### PR DESCRIPTION
**Pull request by kalessil**
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

Reduces amount of code by using generator delegation where possible (language level migration).

### CODE REVIEW

**Review by Jean85 — APPROVED** [review]

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/VariadicValueResolver.php [hunk 1]
```diff
@@ -41,8 +41,6 @@ public function resolve(Request $request, ArgumentMetadata $argument)
             throw new \InvalidArgumentException(sprintf('The action argument "...$%1$s" is required to be an array, the request attribute "%1$s" contains a type of "%2$s" instead.', $argument->getName(), \gettype($values)));
         }
 
-        foreach ($values as $value) {
-            yield $value;
-        }
+        yield from $values;
```
[the remark this answers is not in the document]
Comment (kalessil) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/VariadicValueResolver.php]: @iltar: is the change safe here?

File: src/Symfony/Component/Console/Helper/Table.php [hunk 2]
```diff
@@ -485,9 +485,7 @@ private function buildTableRows($rows)
                 yield $this->fillCells($row);
 
                 if (isset($unmergedRows[$rowKey])) {
-                    foreach ($unmergedRows[$rowKey] as $row) {
-                        yield $row;
-                    }
+                    yield from $unmergedRows[$rowKey];
```
[the remark this answers is not in the document]
Comment (kalessil) [reply src/Symfony/Component/Console/Helper/Table.php]: That's going to be interesting, thank you for the review @stof!

[the remark this answers is not in the document]
Comment (kalessil) [reply src/Symfony/Component/Console/Helper/Table.php]: @ro0NL, wes the loop key was omitted here intentionally, or it can be used for generation delegating (asking as not too deep in the context here)?

[the remark this answers is not in the document]
Comment (ro0NL) [reply src/Symfony/Component/Console/Helper/Table.php]: my change only preserved the current behavior back then... but LGTM.

the internals are a bit hard to follow here (we only use this for `foreach ($rows as $row)` though), but i believe it can cause duplicate keys due the yield above (which shouldnt hurt https://3v4l.org/Yaljq)

if tests pass im :+1:

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/VariadicValueResolver.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (iltar) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/VariadicValueResolver.php]: @kalessil What the code here does is transforming `&arr[]=foo&arr[]=bar` to `['foo', 'bar']` in php, which then get yielded to provide support for `action(...$arr)` I believe. I think @stof made a valid remark and I don't know if this will break anything, as variadic arguments can only receive numeric keys afaik. Would require the addition of a test case to verify if non-numeric keys will work here (which I fear do not).

File: src/Symfony/Component/Console/Helper/Table.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (stof) [reply src/Symfony/Component/Console/Helper/Table.php]: well, as TableRows does not care about the keys (and does not cast it to array, which might cause issues regarding duplicate keys), it is not an issue then.

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/VariadicValueResolver.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (stof) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/VariadicValueResolver.php]: the question is whether the value resolver will keep working fine if `$value` is an associative array (which is totally possible, as this is something coming from outside Symfony itself)

[the remark this answers is not in the document]
Comment (kalessil) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/VariadicValueResolver.php]: Right, `action(...$arr)` requires array to be indexed from 0 to N sequentially. So, the change will be BC if someone provides indexes. `yield from array_values($values);` would be a possible solution.

[the remark this answers is not in the document]
Comment (stof) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/VariadicValueResolver.php]: but that force to first allocate a new array to iterate over it, in which case the existing code is better (it avoids the allocation of the array)

[the remark this answers is not in the document]
Comment (kalessil) [reply src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/VariadicValueResolver.php]: Agree, let's revert the both cases you've mentioned. I obviously need to differentiate case with single value and key-value pair yielding in my plugin.

**Review by nicolas-grekas — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Comment by fabpot** [comment]
Should be for master as it's not a trivial change as proved by the comments (and not a bug fix either).

**Comment by kalessil** [comment]
@ro0NL, the syntax is not supported by PHP prior to 7.0, so only 4.1 and master can be targeted. Later merging can require re-introducing foreach, theoretically.

**Comment by nicolas-grekas** [comment]
Thank you @kalessil.
